aws/resource_aws_appsync_resolver.go

```diff
@@ -0,0 +1,182 @@
+package aws
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/appsync"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsAppsyncResolver() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsAppsyncResolverCreate,
+		Read:   resourceAwsAppsyncResolverRead,
+		Update: resourceAwsAppsyncResolverUpdate,
+		Delete: resourceAwsAppsyncResolverDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"api_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"field": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"data_source": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"request_template": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"response_template": {
+				Type:     schema.TypeString,
+				Required: true, // documentation bug, the api returns 400 if this is not specified.
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsAppsyncResolverCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appsyncconn
+
+	input := &appsync.CreateResolverInput{
+		ApiId:                   aws.String(d.Get("api_id").(string)),
+		DataSourceName:          aws.String(d.Get("data_source").(string)),
+		TypeName:                aws.String(d.Get("type").(string)),
+		FieldName:               aws.String(d.Get("field").(string)),
+		RequestMappingTemplate:  aws.String(d.Get("request_template").(string)),
+		ResponseMappingTemplate: aws.String(d.Get("response_template").(string)),
+	}
+
+	mutexKey := fmt.Sprintf("appsync-schema-%s", d.Get("api_id").(string))
+	awsMutexKV.Lock(mutexKey)
+	defer awsMutexKV.Unlock(mutexKey)
+
+	_, err := retryOnAwsCode(appsync.ErrCodeConcurrentModificationException, func() (interface{}, error) {
+		return conn.CreateResolver(input)
+	})
+
+	if err != nil {
+		return err
+	}
+
+	d.SetId(d.Get("api_id").(string) + "-" + d.Get("type").(string) + "-" + d.Get("field").(string))
+
+	return resourceAwsAppsyncResolverRead(d, meta)
+}
+
+func resourceAwsAppsyncResolverRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appsyncconn
+
+	apiID, typeName, fieldName, err := decodeAppsyncResolverID(d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	input := &appsync.GetResolverInput{
+		ApiId:     aws.String(apiID),
+		TypeName:  aws.String(typeName),
+		FieldName: aws.String(fieldName),
+	}
+
+	resp, err := conn.GetResolver(input)
+	if err != nil {
```

Comment:
Prior to returning an error when an AppSync Resolver cannot be found (which can happen when the GraphQL API or Resolver is deleted outside Terraform), we should instead attempt to catch the error and signal to Terraform that this resource needs to be recreated instead.

```suggestion

	if isAWSErr(err, appsync.ErrCodeNotFoundException, "") {
		log.Printf("[WARN] AppSync Resolver (%s) not found, removing from state", d.Id())
		d.SetId("")
		return nil
	}

	if err != nil {
```

By returning the `appsync.GraphqlApi` and `appsync.Resolver` API objects from the `Exists` testing functions for both resources, this can be tested with the following:

```go
func TestAccAwsAppsyncResolver_disappears(t *testing.T) {
	var api1 appsync.GraphqlApi
	var resolver1 appsync.Resolver
	rName := fmt.Sprintf("tfacctest%d", acctest.RandInt())
	appsyncGraphqlApiResourceName := "aws_appsync_graphql_api.test"
	resourceName := "aws_appsync_resolver.test"

	resource.ParallelTest(t, resource.TestCase{
		PreCheck:     func() { testAccPreCheck(t) },
		Providers:    testAccProviders,
		CheckDestroy: testAccCheckAwsAppsyncResolverDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccAppsyncResolver_basic(rName),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckAwsAppsyncGraphqlApiExists(appsyncGraphqlApiResourceName, &api1),
					testAccCheckAwsAppsyncResolverExists(resourceName, &resolver1),
					testAccCheckAwsAppsyncResolverDisappears(&api1, &resolver1),
				),
				ExpectNonEmptyPlan: true,
			},
		},
	})
}

func testAccCheckAwsAppsyncResolverDisappears(api *appsync.GraphqlApi, resolver *appsync.Resolver) resource.TestCheckFunc {
	return func(s *terraform.State) error {
		conn := testAccProvider.Meta().(*AWSClient).appsyncconn

		input := &appsync.DeleteResolverInput{
			ApiId:     api.ApiId,
			FieldName: resolver.FieldName,
			TypeName:  resolver.TypeName,
		}

		_, err := conn.DeleteResolver(input)

		return err
	}
}
```

Prior to resource code update:

```
--- FAIL: TestAccAwsAppsyncResolver_disappears (12.60s)
    testing.go:538: Step 0 error: Error on follow-up refresh: 1 error occurred:
        	* aws_appsync_resolver.test: 1 error occurred:
        	* aws_appsync_resolver.test: aws_appsync_resolver.test: NotFoundException: No resolver found.
```

After code update:

```
--- PASS: TestAccAwsAppsyncResolver_disappears (17.10s)
```